### ISSUE TITLE
tx.Rollback returns conn busy error as i unable to close BatchResult

### ISSUE THREAD

**Issue by SLIpros** [jackc/pgx#1578]
Hello, calling method `Close` on `BatchResult` entity returns same db error as previously called `Exec` function.
```
new row for relation "promo_codes" violates check constraint "check_promo_code_len" 
``` 
Since I am executing batch request in a transaction after error i am calling `tx.Rollback` that returns error.
```
conn busy
```

Is there any reason we return early and not closing pipeline in BatchResult Close method? What to do in my case and how to properly close `BatchResult`?
```go
// Close closes the batch operation. Any error that occurred during a batch operation may have made it impossible to
// resyncronize the connection with the server. In this case the underlying connection will have been closed.
func (br *pipelineBatchResults) Close() error {
	defer func() {
		if !br.endTraced {
			if br.conn.batchTracer != nil {
				br.conn.batchTracer.TraceBatchEnd(br.ctx, br.conn, TraceBatchEndData{Err: br.err})
			}
			br.endTraced = true
		}
	}()

	if br.err != nil {
		// here is my problem appears. If remove this early return all working fine. 
		// I am getting same error, but pipeline will be closed
		return br.err 
	}
```

**Comment by jackc** [comment]
I just tried to duplicate this but was not successful. Can you provide a reproduction case?

**Comment by SLIpros** [comment]
```go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

const migration = `
	create table promo_codes (
		id UUID PRIMARY KEY,
		code varchar(255) not null,
		len integer not null
			constraint check_promo_code_len check ( len = length(code) )
	);
`

var ErrNothingWasAffected = errors.New("nothing was affected")

func main() {
	dsn := "postgres://postgres:mysecretpassword@192.168.0.100:5555/batch"

	ctx := context.Background()
	conn, err := pgxpool.New(ctx, dsn)
	if err != nil {
		panic("Unable to connect to database: " + err.Error())
	}

	if err := conn.Ping(ctx); err != nil {
		panic("Unable to ping database: " + err.Error())
	}

	defer conn.Close()

	if err := transaction(ctx, conn); err != nil {
		panic("Something bad happens: " + err.Error())
	}

	fmt.Println("Success")
}

func transaction(ctx context.Context, conn *pgxpool.Pool) error {
	tx, err := conn.Begin(ctx)
	if err != nil {
		return err
	}

	defer func() {
		if err := tx.Rollback(ctx); err != nil {
			log.Println("Unable to rollback tx", err)
			return
		}

		log.Println("Successfully rollback tx")
	}()

	if err := batch(ctx, tx); err != nil {
		err := tx.Rollback(ctx)
		if err != nil {
			log.Println("Unable to rollback tx", err)
		}

		return err
	}

	if err = tx.Commit(ctx); err != nil {
		log.Println("Unable to commit tx", err)
	}

	return err
}

func batch(ctx context.Context, tx pgx.Tx) error {
	id := "0187989f-4aed-729e-bc80-b3feed1a1b9c"
	promocode := "79037994250"
	promocodeLen := 14 // wrong len, if switch it to 11 there would be no errors

	var b pgx.Batch
	b.Queue("INSERT INTO promo_codes(id, code, len) VALUES ($1, $2, $3)", id, promocode, promocodeLen)

	result := tx.SendBatch(ctx, &b)
	defer func() {
		if err := result.Close(); err != nil {
			log.Println("Unable to close batch result", err)
			return
		}

		log.Println("Successfully close batch result")
	}()

	for i := 0; i < b.Len(); i++ {
		tag, err := result.Exec()
		if err != nil {
			log.Println("Exec error", err)

			return err
		}

		if tag.RowsAffected() == 0 {
			return ErrNothingWasAffected
		}
	}

	return nil
}
```

Output
```
2023/04/19 12:45:21 Exec error ERROR: new row for relation "promo_codes" violates check constraint "check_promo_code_len" (SQLSTATE 23514)
2023/04/19 12:45:21 Unable to close batch result ERROR: new row for relation "promo_codes" violates check constraint "check_promo_code_len" (SQLSTATE 23514)
2023/04/19 12:45:21 Unable to rollback tx conn busy
2023/04/19 12:45:21 Unable to rollback tx tx is closed
panic: Something bad happens: conn busy
```

**Comment by jackc** [comment]
The problem was a combination of two issues.

1. The batch strategy using pipeline mode was modeled after the original strategy that did not. In that original implementation (that is still used for some queries) the cleanup happened when an error was encountered so the Close method being a no-op was correct. With pipeline mode that is not the case.
2. This error was not caught by tests because the tests are using the newer callback style for reading results with `QueuedQuery`. But this bug only occurred if the results were read from `BatchResults` directly. 

Should be fixed as of 09371981f92f113f676010458462a3b0ebc38c1f.

However, you may want to consider using the callback functions (e.g. `batch.Queue("select ...").Query(...)`) as it is usually easier to understand when the query and the code that handles the result are next to each other.
